docs/mfc/tn035-using-multiple-resource-files-and-header-files-with-visual-cpp.md

```diff
@@ -47,17 +47,22 @@ This note is structured to answer the above questions as follows:
+#### Read-Only Symbol Directives
 
-### Compile-Time Directives
+Although Visual C++ only edits one header file for any given .RC file, Visual C++ supports references to symbols defined in additional read-only header files. The **Read-Only symbol Directives:** section in the **Resource Includes** dialog in Visual C++, lets you specify any number of additional read-only header files as Read-Only Symbol Directives. The "read-only" restriction means that when you add a new resource in the .RC file, you can use a symbol defined in the read-only header file; but if you delete the resource, the symbol still remains defined in the read-only header file. You cannot change the numeric value assigned to a read-only symbol.
+
+#### Compile-Time Directives
 
-Visual C++ also supports nesting of resource files, where one .RC file is #include'd within another. When you edit a given .RC file using Visual C++, any resources in the #include'd files are not visible. But when you compile the .RC file, the #include'd files are also compiled. Using the **Resource Includes** command from the **View** menu in Visual C++, you can specify any number of #include'd .RC files as Compile-Time Directives.
+Visual C++ also supports nesting of resource files, where one .RC file is #include'd within another. When you edit a given .RC file using Visual C++, any resources in the #include'd files are not visible. But when you compile the .RC file, the #include'd files are also compiled. The **Compile-Time Directives:** section in the **Resources Includes** dialog in Visual C++, lets you specify any number of #include'd .RC files as Compile-Time Directives.
 
 Note what happens if you read into Visual C++ an .RC file that #include's another .RC file that is *not* specified as a Compile-Time Directive. This situation might arise when you bring to Visual C++ an .RC file that you had been previously maintaining manually with a text editor. When Visual C++ reads the #include'd .RC file, it merges the #include'd resources into the parent .RC file. When you save the parent .RC file, the #include statement, in effect, will be replaced by the #include'd resources. If you do not want this merge to happen, you should remove the #include statement from the parent .RC file *prior* to reading it into Visual C++; then using Visual C++, add back the same #include statement as a Compile-Time Directive.
```

Comment:
Note what happens if you read into Visual C++ an .RC file that `#include`'s another .RC file that is *not* specified as a Compile-Time Directive. This situation might arise when you bring to Visual C++ an .RC file that you had been previously maintaining manually with a text editor. When Visual C++ reads the `#include`'d .RC file, it merges the `#include`'d resources into the parent .RC file. When you save the parent .RC file, the `#include` statement, in effect, will be replaced by the `#include`'d resources. If you do not want this merge to happen, you should remove the `#include` statement from the parent .RC file *prior* to reading it into Visual C++; then using Visual C++, add back the same `#include` statement as a Compile-Time Directive.

Reply:
How about:
At compile time, Visual Studio merges `#include` statements in a child `.RC` file into the parent `.RC` file that `#include`s it. If you don't want this behavior, remove the `#include` statement from the parent `.RC` file *prior* to adding the file to Visual Studio. Then, in Visual Studio, add the `#include` statement as a **Compile-Time Directive**.

Even with this change to try to get to the heart of the matter, I'm still not sure what exactly happens. Are you saying that VS rewrites the parent .RC file to have all the #include statements from the child?

Reply:
> ...it merges the #include'd resources into the parent .RC file.

How about:
Before compile time, resources like *Ribbon, Menu, Bitmap, Dialog*, etc. are visually edited resulting in changes to, for example, screen positions and IDs and target objects. The `ribbon.mfcribbon-ms` xml file organizes ID, file name changes and object relations for the ribbon. 

At compile time, which occurs when the Visual Studio program is built, these changed resource files become RC Compiler input and get compiled with other **Compile-time directives** like `afxres.rc` and `afxprint.rc` as well as **Read-only symbol directives** like `resource.h` and `myResource.h`. This RC Compiling produces a new `.rc` file. It is this *.rc* file that serves as object data for the rest of the build.

Reply:
Hi @rtischer8277 ,  now we seem to have lost the suggestion you were making to devs: "If you do not want this merge to happen..."  Should that be part of this?  You can also remove redundancy such as explaining what compile time is.

Reply:
Are you on-board with my reasoning for this section?

Reply:
This is too big for a note. It should be its own section. It's also hard to scan.  Can it be broken down into bullets or a table for special cases and the corresponding resolution?

What does this mean: If that occurs, a previous .rc file will be needed to bring RC Compiling back to a functioning state

Reply:
@rtischer8277, thank you for hanging in on these changes. I think you are making an important contribution to the article.
Once we get through the main ideas of your changes, I think rather than do an editing pass back and forth in this PR, I'll have you check in and later I'll go through and do the all the Microsoft editing things like using correct branding, style updates, etc. Good with you?

Reply:
That's what I proposed, I thought :-)

Reply:
I would like to take a stab at re-writing the *"Note what happens if you read into Visual C++ an .RC file..."* paragraph. I have collected all the edits and consolidated them in the following text:

### **Resource Includes** directives ( non-primary `.rc` script file(s) not added to Visual Studio project)
> [!NOTE]
You can access the **Resource Includes** by going to the **Resource View** then right clicking the `.rc` file and selecting **Resource Includes**.

####Symbol Header File
By default, Visual C++ always names the corresponding header file `RESOURCE.H`, regardless of the name of the resource file (e.g., `MYAPP.RC`). Using the **Resource Includes** command from the **Resource View** menu in Visual C++ lets you can change the name of this header file by updating the **Symbol Header File** file name in the **Set Includes** edit box.
> [!NOTE]
If the header file isn't in the same directory as the project, in **Symbol Header File**, prepend the new `MYRESOURCE.H` header name with a relative path.

####**Read-only symbol directives**
Although Visual C++ only edits one header file for any given .RC file, Visual C++ supports references to symbols defined in additional read-only header files. The **Read-only symbol directives** section in the **Resource Includes** dialog in Visual C++ lets you specify any number of additional header files.

####**Compile-time directives**
Other non-primary `.rc` script files may be added directly to your primary `.rc` script file using this section. Their effect is the following:

Before compile time, resources like *Ribbon, Menu, Bitmap, Dialog*, etc. are visually edited resulting in changes to, for example, screen positions and IDs and target objects. While the `.rc` file organizes IDs, file name changes and their paths, the `ribbon.mfcribbon-ms` xml file organizes IDs and object relations to each other for the ribbon. Resource editing changes affect this file.

At compile time, which occurs when the Visual Studio program is built, these changed resource files become `RC` Compiler input and get compiled with other **Compile-time directives** like `afxres.rc` and `afxprint.rc` as well as **Read-only symbol directives** like `resource.h` and `myResource.h`. This `RC` Compiling produces a new `.rc` file. It is this `.rc` file that serves as object data for the rest of the build.

Visual Studio C++ compiles output a new `.rc` file reflecting the three kinds of **Resource Includes** information (**Symbol header file**, **Read-only symbol directives**, and **Compile-time directives**). `#include` directives appear in `TEXTINCLUDE` resources. The `TEXTINCLUDE` resources, an implementation detail that you do not normally need to deal with, are explained in [How Visual C++ Manages Set Includes Information](https://docs.microsoft.com/en-us/cpp/mfc/tn035-using-multiple-resource-files-and-header-files-with-visual-cpp?view=msvc-170#_mfcnotes_tn035_set_includes).

### **Resource Includes** directives (non-primary `.rc` script file(s) added to Visual Studio project)
Another way to deal with other `.rc` script files is to add them directly to your Visual Studio project's solution. Your other `.rc` files will then appear as separate menu items in your solution's *Resource View*. One advantage to this method is Visual Studio manages access to the file's location, so there is no need for relative paths. Another advantage is, you can use any or all of Visual Studio's resource editors to manipulate its resource contents. There is, however, a major disadvantage. When using the resource editors to make changes in a non-primary `.rc` file, the `RC` Compiler is not able to save the changes and shows the *Cannot save file* message. Once visual `.rc` script editing is in this state, it is necessary to close the Visual Studio project and reopen it in order to discard the changes. To not get in this resource editing state, do all your resource editing in your main `.rc` file which is always able to save your changes. Once the visual editing changes are complete, save them, and then copy manually any changed `.rc` script definition blocks into your destination `.rc` file(s). The sum of the script definitions must, however, always be mutually exclusive at build time. To achieve that, temporarily comment out the definition blocks in the non-primary `.rc` files, that are duplicates.

Reply:
I like this version. I think it is only missing some context about why you want to read this note. I'm thinking of an introduction along the lines of "you want to do x, here's how" or "you have problem x, here's how" or "you've wondered how this machinery works in detail, here's how" that sets up why we are reading this text.

Reply:
I saw Anju's comment, but didn't understand it. Maybe with more explanation I could see where it may apply to the current draft. I suggest that the comment be applied to another PR in the future where it can be fully discussed.

I agree with your context intro and here is some introductory text. If the text is too wordy, let me know and I'll cut it down, or you can. I am also setting *Conversation Resolved* and am ready to check it in:

I suggest that we proceed towards check-in with this PR.

Here is the suggested Introduction to the draft:

**Introduction to cross-project `.rc` script file re-use**
What marks production code from experimental code is the necessity for eliminating code duplication. To the degree present, duplicate code renders production code project(s) un-maintainable over time. To combat this tendency, C++ has always observed the one-definition rule (ODR) and no changes are allowed into ISO C++ that do not adhere to ODR. MFC C++ can be written to observe ODR as well, and therefore can maintain cross-project maintainability thanks to ISO C++. But Visual Studio C++'s `.rc` script file is not C++ and therefore requires special attention to achieve cross-project ODR status. This means breaking up `.rc` files into separate scripts that make sense for your set of projects. Below is a description of how this is done:

Reply:
Thanks, @rtischer8277 . I pinged Anju to respond to the issue he raised. I think the ISO C++ connection is unnecessary (and a slightly stretched connection in this context). People will understand that you don't want duplicate copies of stuff. I'd just say you don't want duplicate copies of things in your resources, here's how to do it.

Reply:
Great